### PR TITLE
Dellvmize some intrinsics (use `u32` instead of `Self` in some integer intrinsics)

### PR DESCRIPTION
**Pull request by WaffleLapkin**
This implements https://github.com/rust-lang/compiler-team/issues/693 minus what was implemented in #123226.

Note: I decided to _not_ change `shl`/... builder methods, as it just doesn't seem worth it.

r? @scottmcm

### CODE REVIEW

File: compiler/rustc_const_eval/src/interpret/intrinsics.rs [hunk 1]
```diff
@@ -487,7 +492,7 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
             sym::bitreverse => (bits << extra).reverse_bits(),
             _ => bug!("not a numeric intrinsic: {}", name),
         };
-        Ok(Scalar::from_uint(bits_out, layout.size))
+        Ok(Scalar::from_uint(bits_out, ret_layout.size))
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/intrinsics.rs:495]: This looks like it can ICE when bswap/bitrevere is called with a too small ret_layout.

**Review by scottmcm — CHANGES_REQUESTED** [review]
(Looks like some CI issues?)

**Comment by WaffleLapkin** [comment]
`#[cfg(not(bootstrap))]` makes tidy think unsafe is not documented. fun.

File: compiler/rustc_const_eval/src/interpret/intrinsics.rs (hunk 1 above)
Comment (WaffleLapkin) [reply compiler/rustc_const_eval/src/interpret/intrinsics.rs:495]: How would that happen though? The intrinsics are defined like `pub fn bitreverse<T: Copy>(x: T) -> T;`, so the `ret_layout` must be the same?

**Comment by WaffleLapkin** [comment]
For some reason any binary I compile with this branch gets a breakpoint in `std::panicking::r#try` (even if it's just an empty main!):

```
(gdb) run
Starting program: /home/gh-WaffleLapkin/rust-tsk/t 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/aarch64-linux-gnu/libthread_db.so.1".

Program received signal SIGTRAP, Trace/breakpoint trap.
std::panicking::try<(), std::rt::lang_start_internal::{closure_env#1}> () at library/std/src/panicking.rs:516
516             return if intrinsics::catch_unwind(do_call::<F, R>, data_ptr, do_catch::<F, R>) == 0 {
(gdb) bt
#0  std::panicking::try<(), std::rt::lang_start_internal::{closure_env#1}> () at library/std/src/panicking.rs:516
#1  0x0000aaaaaaaac308 in std::panic::catch_unwind<std::rt::lang_start_internal::{closure_env#1}, ()> () at library/std/src/panic.rs:149
#2  std::rt::lang_start_internal () at library/std/src/rt.rs:140
#3  0x0000aaaaaaaa4e5c in std::rt::lang_start ()
#4  0x0000aaaaaaaa4f44 in main ()
(gdb) disassemble
Dump of assembler code for function _ZN3std9panicking3try17h0087a381859a0704E:
   0x0000aaaaaaaaad6c <+0>:     stp     x29, x30, [sp, #-16]!
   0x0000aaaaaaaaad70 <+4>:     mov     x29, sp
=> 0x0000aaaaaaaaad74 <+8>:     brk     #0x1
End of assembler dump.
```

Clearly the function is miscompiled, but what could have made it miscompile?... I did not touch anything relating to it...

File: compiler/rustc_const_eval/src/interpret/intrinsics.rs (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/intrinsics.rs:495]: That's an extremely non-local invariant. This function here should make sense on its own without worrying about what happens in the standard library.

Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/intrinsics.rs:495]: Maybe add an assertion that for bscap/bitrevers, the two layouts are the same.

Comment (WaffleLapkin) [reply compiler/rustc_const_eval/src/interpret/intrinsics.rs:501]: I think this is true in general with intrinsics, we implicitly depend on their signature in a lot of ways. Either way, I added an assertion.

File: compiler/rustc_codegen_llvm/src/intrinsic.rs [hunk 2]
```diff
@@ -355,6 +355,10 @@ impl<'ll, 'tcx> IntrinsicCallMethods<'tcx> for Builder<'_, 'll, 'tcx> {
                             // rotate = funnel shift with first two args the same
                             let llvm_name =
                                 &format!("llvm.fsh{}.i{}", if is_left { 'l' } else { 'r' }, width);
+
+                            // llvm expects shift to be the same type as the values, but rust always uses `u32`
+                            let raw_shift = self.intcast(raw_shift, self.val_ty(val), false);
```
[the remark this answers is not in the document]
Comment (WaffleLapkin) [reply compiler/rustc_codegen_llvm/src/intrinsic.rs:367]: Done

File: compiler/rustc_const_eval/src/interpret/intrinsics.rs (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/intrinsics.rs:501]: In the interpreter and Miri we always have local assertions for these things.

**Comment by WaffleLapkin** [comment]
@rustbot ready

I was not able to fix the gcc backend (could not run tests locally and using CI seemed like too much trouble), so I just commented the tests out ^^'

Other backends should be working, unless I missed something.

**Comment by antoyo** [comment]
The error we see above for the GCC codegen is in the frontend. That should be fixed by doing the same change that was done for cranelift [here](https://github.com/rust-lang/rust/pull/124003/files#diff-66b7cd69caf1788ec7feab64ad0fd9171efdbd4a23633f15b5f8fc7496efc8f5L630).

**Comment by antoyo** [comment]
Is the error you got from the GCC codegen the one we see in the message above?
This one:
```
##[error]   --> example/example.rs:157:5
    |
    |
156 | unsafe fn use_ctlz_nonzero(a: u16) -> u16 {
    |                                       --- expected `u16` because of return type
157 |     intrinsics::ctlz_nonzero(a)
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `u16`, found `u32`
    |
help: you can convert a `u32` to a `u16` and panic if the converted value doesn't fit
    |
157 |     intrinsics::ctlz_nonzero(a).try_into().unwrap()
```
?
Because that's a frontend error, so that seems weird to me especially since you changed the signature [here](https://github.com/rust-lang/rust/pull/124003/files#diff-91388813554b357988af6fe24ec77979c00708f4c5d2cdb9c22f1116ba6f6d7bL596).

**Comment by WaffleLapkin** [comment]
@antoyo this is the error I got, but if I'd fix it, I would just get a different error, because gcc (now) casts to the wrong type

**Comment by antoyo** [comment]
Did you try to change `arg_type` to `self.u32_type` in [this code](https://github.com/rust-lang/rustc_codegen_gcc/blob/master/src/intrinsic/mod.rs#L718-L773)?

**Comment by WaffleLapkin** [comment]
I did not, precisely because I wasn't able to test things

**Comment by antoyo** [comment]
Ok, you might want to join [this discussion](https://github.com/rust-lang/rust/issues/124172) so we can figure out how to make this better.

**Review by scottmcm — APPROVED** [review]
The changes here all look good to me.

@RalfJung , did you want to look further at the CTFE parts?

I'm a bit uncertain about the cg_gcc state here.  The amount being disabled is small so seems fine enough to do and resolve later, but I also don't feel confident in saying what the expectations are for it.

**Comment by RalfJung** [comment]
The interpreter changes LGTM.

**Comment by antoyo** [comment]
> I'm a bit uncertain about the cg_gcc state here. The amount being disabled is small so seems fine enough to do and resolve later, but I also don't feel confident in saying what the expectations are for it.

While we do prefer to avoid disabling code like was done here (and we're always happy to help implement the stuff in the PR), we're OK with people not implementing the stuff and doing it on our side.

I guess we could think at some point of allowing us to fix the cg_gcc stuff in PR made by other people, but we should also think about that carefully to not block people.

All of that to say, I'm OK with the cg_gcc changes here.

**Comment by workingjubilee** [comment]
hokay!

@bors r=scottmcm,RalfJung,antoyo

**Comment by WaffleLapkin** [comment]
Thanks to everyone involved <3

**Pull request merged**